src/sentry/auth/access.py

```diff
@@ -179,47 +211,58 @@ def has_team_scope(self, team: Team, scope: str) -> bool:
         if self.has_scope(scope):
             return True
 
-        membership = self._team_memberships.get(team)
-        if membership is not None and scope in membership.get_scopes():
-            metrics.incr(
-                "team_roles.pass_by_team_scope",
-                tags={"team_role": membership.role, "scope": scope},
-            )
-            return True
+        for member in self.api_member.teams:
+            if member.team.id == team.id:
+                if scope in member.scopes:
+                    metrics.incr(
+                        "team_roles.pass_by_team_scope",
+                        tags={"team_role": member.role, "scope": scope},
+                    )
+                    return True
+                else:
+                    break
 
         return False
 
-    def has_team_membership(self, team: Team) -> bool:
-        return team in self.teams
+    def has_team_membership(self, team: Union[Team, ApiTeam]) -> bool:
+        for t in self.api_team_members:
+            if t.team.id == team.id:
+                return True
+        return False
 
     def get_team_role(self, team: Team) -> Optional[TeamRole]:
-        team_member = self._team_memberships.get(team)
-        return team_member and team_member.get_team_role()
+        for t in self.api_team_members:
```

Comment:
Should `api_team_members` be stored as dictionary? We need to do a bunch of loops on it to find specific teams.

Reply:
Possibly?  I've been trying to avoid a complex interface, however, in part because the existing access was structured in such a way that made it kind of hard to refactor (tons of similar, but not quite the same, interfaces to data.  In fact the existing `teams` interface was really annoying to unpack.).  I don't think this is a true performance issue because the scale of api_team_members isn't likely to exceed 1000, and honestly a loop with single comparison is pretty damn fast in all cases.

My intuition here was that a simpler interface was fair for the performance trade off given that these methods are, in practice, not called in N * M complexity (constant amount of invocations per endpoint).

Reply:
I'll be continuing down the path of a simpler interface here because I'm going to take a first pass where I *reduce* the access public interface quite a bit (without changing underlying implementation) in one PR before I introduce the organization service changes into it as well.